kratos/solving_strategies/convergencecriterias/mixed_generic_criteria.h

```diff
@@ -0,0 +1,441 @@
+     * @return std::unordered_map<KeyType, KeyType>& Reference to the local key map
+     */
+    std::unordered_map<KeyType, KeyType>& GetLocalKeyMap()
+    {
+        return mLocalKeyMap;
+    }
+
+    /**
+     * @brief Calculate the convergence norms
+     * This method calculates the convergence norms for all the variables to be checked
+     * @param rModelPart Reference to the ModelPart containing the fluid problem.
+     * @param rDofSet Reference to the container of the problem's degrees of freedom (stored by the BuilderAndSolver)
+     * @param rDx Vector of results (variations on nodal variables)
+     * @return std::tuple<std::vector<TDataType>, std::vector<TDataType>> Tuple containing the absolute and relative convergence values
+     */
+    std::tuple<std::vector<TDataType>, std::vector<TDataType>> CalculateConvergenceNorms(
+        const ModelPart& rModelPart,
+        const DofsArrayType& rDofSet,
+        const TSystemVectorType& rDx)
+    {
+        // Initialize
+        std::vector<int> dofs_count(mVariableSize, 0);
+        std::vector<TDataType> solution_norms_vector(mVariableSize, 0.0);
+        std::vector<TDataType> increase_norms_vector(mVariableSize, 0.0);
+
+        // Accumulate the norm values
+        GetNormValues(rModelPart, rDofSet, rDx, dofs_count, solution_norms_vector, increase_norms_vector);
+
+        // Synchronize the norm values
+        const auto& r_data_comm = rModelPart.GetCommunicator().GetDataCommunicator();
+        auto global_solution_norms_vector = r_data_comm.SumAll(solution_norms_vector);
+        auto global_increase_norms_vector = r_data_comm.SumAll(increase_norms_vector);
+        auto global_dofs_count = r_data_comm.SumAll(dofs_count);
+
+        // Check division by zero in global solution norms
+        const double zero_tol = 1.0e-12;
+        for(int i = 0; i < mVariableSize; i++) {
+            if (global_solution_norms_vector[i] < zero_tol) {
+                global_solution_norms_vector[i] = 1.0;
+            }
+        }
+
+        // Calculate the norm values
+        std::vector<TDataType> var_ratio(mVariableSize, 0.0);
+        std::vector<TDataType> var_abs(mVariableSize, 0.0);
+        for(int i = 0; i < mVariableSize; i++) {
+            var_ratio[i] = std::sqrt(global_increase_norms_vector[i] / global_solution_norms_vector[i]);
+            var_abs[i] = std::sqrt(global_increase_norms_vector[i]) / static_cast<TDataType>(global_dofs_count[i]);
+        }
+
+        // Output the ratio and absolute norms as a tuple
+        return std::make_tuple(var_ratio, var_abs);
+    }
+
+    /**
+     * @brief Method to output the convergence status
+     * This method prints the convergence status to the screen for each one of the checked variables
+     * @param rConvergenceNorms Tuple containing the absolute and relative convergence values
+     */
+    void OutputConvergenceStatus(const std::tuple<std::vector<TDataType>, std::vector<TDataType>>& rConvergenceNorms)
```

Comment:
```suggestion
    void OutputConvergenceStatus(const std::tuple<std::vector<TDataType>, std::vector<TDataType>>& rConvergenceNorms) const
```

Reply:
Impossible because of the `operator[]` of the map.